remote-api/src/main/java/org/alfresco/rest/api/search/impl/SearchMapper.java

```diff
@@ -581,9 +581,11 @@ public void fromPivot(SearchParameters sp, List<StatsRequestParameters> stats, F
     {
         if (multiplePivots != null && !multiplePivots.isEmpty())
         {
+            List<FacetField> facets = new ArrayList<>();
+            facets.addAll(facetFields.getFacets());
             multiplePivots.forEach(aPivot -> {
                 List<String> pivotKeys = new ArrayList<>();
-                buildPivotKeys(pivotKeys, aPivot, stats, facetFields, ranges, searchRequestContext);
+                buildPivotKeys(pivotKeys, aPivot, stats, new FacetFields(facets), ranges, searchRequestContext);
```

Comment:
Hi @antoniojfelix I didn't get the point of the change in this PR. 

The previous code uses a give `FacetFields` reference, which is basically a container of a list of `FacetField` instances and passes it to:

- fromPivot  
- fromFacetFields

```
fromPivot(..., searchQuery.getFacetFields(), ...);
...     
fromFacetFields(sp, searchQuery.getFacetFields());
```

The `FacetFields` instance is passed by reference. If pivots are requested and a facet field is requested within the pivots, the `fromPivot` and `buildPivotKeys` methods remove that field from the internal facet fields list in `FacetFields`. 

In this way, we cannot have, at the same time, a field used as a top-level facet field and pivot field. 

Now, the change in the PR, if I got it correctly

- passes to the `buildPivotKeys` a new fresh instance of `FacetFields`
- the new instance is populated with the list (better, the reference to the list) taken from the original `FacetFields`

Doesn't that mean the behavior is still the same? The list is copied by reference; If the it is emptied in `buildPivotKeys`, the original `FacetFields` (subsequently passed to the `fromFacetFields(...)` will still get an empty list.

On top of that, I don't understand how that change relates to the original request. The limit parameter is still not managed in the pivot methods.

Again, maybe I misunderstood something, could you please clarify for me?

Thanks

Reply:
Hello @agazzarini.

I think you understood the basic idea of the change.

When the process runs the `fromPivot(..., searchQuery.getFacetFields(), ...);`,  the `fromPivot` and `buildPivotKeys` methods remove all the fields from the internal facet fields list in FacetFields, exactly like you said.

So, when the `fromFacetFields(sp, searchQuery.getFacetFields());` is called, the list of FacetFields is empty, so the facet fields are not considered and the code `ff.setLimitOrNull(facet.getLimit());` is not called.

To avoid this, I added the `new FacetFields(facets)` as an argument in the `buildPivotKeys` function, and while doing so, this new instance would be emptied instead of the original list. 
Now, the `fromFacetFields(sp, searchQuery.getFacetFields());` will have the list of facet fields to be added to the query, and the code `ff.setLimitOrNull(facet.getLimit());` is now called to add the `facet.limit` to the query as well.

This was the main idea, but I'm not sure that this is the right approach.

While I was doing this, it felt that I was changing the code to add a new feature instead of fixing a bug, and since this was my first search issue, I wanted to confirm with someone from the search team before moving forward. 

 If you don't agree with the fix or/and if you have another solution, feel free to close this PR.

If you have anymore questions, feel free to ask.

Regards

Reply:
Hi @antoniojfelix thanks for the explanation. That's clear.
I think, following what @killerboot said, the weak part is that it sounds like we are adding a feature, meaning that if the list is not cleared, Search Service receives a request with the same field used both as a top-level and nested facet.

So ideally, in order to cover all possible scenarios, we should be able to    

1. request F11 as top-level facet field
2. request F1 as a top-level field and then F11 as nested facet field (+ corresponding parameters) 
3. request F1 and F11 as top-level facet fields 

The crucial point is the 2nd because the current implementation doesn't allow to send facet field parameters without requesting the field itself (as a top-level facet).

Let's see what we can do

Thanks again, 
Andrea